lib/srv/server/ssm_install.go

```diff
@@ -104,6 +151,107 @@ func (si *SSMInstaller) Run(ctx context.Context, req SSMRunRequest) error {
 	return trace.Wrap(g.Wait())
 }
 
+func invalidSSMInstanceEvent(accountID, region, instanceID, status string) apievents.SSMRun {
+	return apievents.SSMRun{
+		Metadata: apievents.Metadata{
+			Type: libevents.SSMRunEvent,
+			Code: libevents.SSMRunFailCode,
+		},
+		CommandID:  "no-command",
+		AccountID:  accountID,
+		Region:     region,
+		ExitCode:   -1,
+		InstanceID: instanceID,
+		Status:     status,
+	}
+}
+
+func (si *SSMInstaller) emitInvalidInstanceEvents(ctx context.Context, req SSMRunRequest, instanceIDsState *instanceIDsSSMState) error {
+	var errs []error
+	for _, instanceID := range instanceIDsState.missing {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"EC2 Instance is not registered in SSM. Make sure that the instance has AmazonSSMManagedInstanceCore policy assigned.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			errs = append(errs, trace.Wrap(err))
+		}
+	}
+
+	for _, instanceID := range instanceIDsState.connectionLost {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"SSM Agent in EC2 Instance is not connecting to SSM Service. Restart or reinstall the SSM service. See https://docs.aws.amazon.com/systems-manager/latest/userguide/ami-preinstalled-agent.html#verify-ssm-agent-status for more details.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			errs = append(errs, trace.Wrap(err))
+		}
+	}
+
+	for _, instanceID := range instanceIDsState.unsupportedOS {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"EC2 instance is running an unsupported Operating System. Only Linux is supported.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			errs = append(errs, trace.Wrap(err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// instanceIDsSSMState contains a list of EC2 Instance IDs for a given state.
+type instanceIDsSSMState struct {
+	valid          []string
+	missing        []string
+	connectionLost []string
+	unsupportedOS  []string
+}
+
+// describeSSMAgentState returns the instanceIDsSSMState for all the instances.
+func (si *SSMInstaller) describeSSMAgentState(ctx context.Context, req SSMRunRequest, allInstanceIDs []string) (*instanceIDsSSMState, error) {
+	ret := &instanceIDsSSMState{}
+
+	const defaultMaxResults = 10
+
+	ssmInstancesInfo, err := req.SSM.DescribeInstanceInformationWithContext(ctx, &ssm.DescribeInstanceInformationInput{
+		Filters: []*ssm.InstanceInformationStringFilter{
+			{Key: aws.String(ssm.InstanceInformationFilterKeyInstanceIds), Values: aws.StringSlice(allInstanceIDs)},
+		},
+		// AWS returns an error if MaxResults is less than 5.
+		MaxResults: aws.Int64(max(defaultMaxResults, int64(len(allInstanceIDs)))),
```

Comment:
The max is 50 though. I think we should just use the default and iterate through NextToken

Reply:
I'll use `awsEC2APIChunkSize` instead and add a comment explaining why we shouldn't change its value.